### ISSUE TITLE
Build grpcio wheel from source on Windows 

### ISSUE THREAD

**Issue by johnwog** [grpc/grpc#19999]
### What version of gRPC and what language are you using?
v1.23.xx

### What operating system (Linux, Windows,...) and version?
Windows 10

### What runtime / compiler are you using (e.g. python version or version of gcc)
Python 3.6 virtual env, MSVC 2017

### What did you do?
I tried to build a python wheel for grpcio with the following flags
set GRPC_PYTHON_BUILD_WITH_CYTHON=1 
set GRPC_PYTHON_USE_PREBUILT_GRPC_CORE=1
set GRPC_PYTHON_BUILD_EXT_COMPILER_JOBS=2

gRPC core was built according to: 
grpc/test/distrib/cpp/run_distrib_test_cmake.bat

gRPC wheel was built according to:
grpc/tools/run_tests/artifacts/build_artifact_python.bat
If possible, provide a recipe for reproducing the error. Try being specific and include code snippets if helpful.

It fails on the first build_ext command
python setup.py build_ext --inplace -vvv
 (setup.py at the root dir of grpc repo)

### What did you expect to see?
a compiled wheel

### What did you see instead?
running build_ext
error: [WinError 2] The system cannot find the file specified

if I run with verobose this is part of the stdout
import 'Cython.Compiler.CythonScope' # <_frozen_importlib_external.SourceFileLoader object at 0x00000206C935BE10>
credentials_dir src\python\grpcio\grpc\_cython\_credentials
PACKAGES ['grpc', 'grpc.beta', 'grpc.experimental', 'grpc.framework', 'grpc._cython', 'grpc.framework.common', 'grpc.framework.foundation', 'grpc.framework.interfaces', 'grpc.framework.interfaces.base', 'grpc.framework.interfaces.face', 'grpc._cython._cygrpc']
CYTHON_EXTENSION_MODULES [<setuptools.extension.Extension('grpc._cython.cygrpc') at 0x206c9f95390>]
PACKAGE_DIRECTORIES {'': 'src\\python\\grpcio'}
INSTALL_REQUIRES {'': 'src\\python\\grpcio'}
SETUP_REQUIRES {'': 'src\\python\\grpcio'}
EXTRA_ENV_COMPILE_ARGS  -std=c++11 -DPB_FIELD_32BIT
CORE_C_FILES <filter object at 0x00000206C7E9C940>
running build_ext
 C:\ProgramData\Anaconda3\envs\grpcio\lib\distutils\command\__pycache__\build.cpython-36.pyc matches C:\ProgramData\Anaconda3\envs\grpcio\lib\distutils\command\build.py
 code object from 'C:\\ProgramData\\Anaconda3\\envs\\grpcio\\lib\\distutils\\command\\__pycache__\\build.cpython-36.pyc'
import 'distutils.command.build' # <_frozen_importlib_external.SourceFileLoader object at 0x00000206CA1076D8>
error: [WinError 2] The system cannot find the file specified

### Anything else we should know about your project / environment?
I suspect the issue has to do with relative and absolute paths but I have no idea where to fix that. The setup scripts are quite complicated. Reading around here and on stackoverflow people had issues with the build in the past. The root was a setuptools bug that was supposed to be resolved and installing the latest version did the trick. I reinstalled the latest setuptools and tried also with older versions but the error is still there. 

I read your README where it's stated 
"You cannot currently install Python from source on Windows. Things might work out for you in MSYS2 (follow the Linux instructions), but it isn't officially supported at the moment".

How do you build the wheel for windows you distribute on pypi then?

Any help is greatly appreciated.

**Comment by narenandu** [comment]
Same here...  Failing to build a `whl` file for `grpcio` on Windows. Any pointers would be really helpful !

```
(venv) PS C:\grpc> python setup.py build
Cython-generated files are missing...
running build
running build_py
running build_project_metadata
copying src\python\grpcio\grpc\_grpcio_metadata.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_cython\_credentials\roots.pem -> python_build\lib.win-amd64-2.7\grpc\_cython\_credentials
warning: BuildPy: byte-compiling is disabled, skipping.

running build_ext
error: [Error 2] The system cannot find the file specified
```

**Comment by gnossen** [comment]
@narenandu What version of gRPC are you using? There should have been [a very recent fix](https://github.com/grpc/grpc/pull/21820) for this.

**Comment by gnossen** [comment]
Speaking of which, this is a duplicate of https://github.com/grpc/grpc/issues/20879 and should already be resolved. Please comment if you're still having troubles after updating to master.

**Comment by narenandu** [comment]
@gnossen 
**TL;DR version**
`1.21.1` 

----
**Longer Version**

we have a specific usecase of building `gprcio` `whl` with a custom python interpreter (More details about the python interpreter here: https://clamdragon3d.com/blog/2018/3/22/python-virtualenvs-and-maya). There is an issue of `Stream Removed`, that prompted us to check for the `grpcio` version which fixed the issue with `tcp_windows.cc` file and thus we ended up figuring out that `grpcio==1.21.1` is the minimum version necessary to fix the issue.
```
_Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNKNOWN
	details = "Stream removed"
	debug_error_string = "{"created":"@1573001562.066000000","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"Stream removed","grpc_status":2}"
```
So trying to build the python library by checking out `v1.21.1` in the repository with the custom python interpreter

**Comment by lidizheng** [comment]
@narenandu Is it possible to use our latest version? We only provide support for last two major releases.

A work around can be manually patching https://github.com/grpc/grpc/pull/21820 into your clone.

**Comment by gnossen** [comment]
I'm not sure I understand the blocker to using `master`. Are you saying that the "stream removed" issue is present on `master` and the latest released version that does *not* exhibit this issue is `1.21.1`?

**Comment by narenandu** [comment]
There is no problem in jumping to `master`, thought I would upgrade only to the minimum version that resolves our issue. We are on `1.16.1` and `1.21.1` resolves the issue. But I can definitely jump to the latest available version and try to build it, if that generates the `whl` file needed for my purpose.

Will update on the progress. 

Thanks

**Comment by gnossen** [comment]
> I would upgrade only to the minimum version that resolves our issue

Why? As @lidizheng pointed out, we only support the last two releases. I quickly scanned the article you linked and didn't see a justification for this.

**Comment by narenandu** [comment]
Ok. I tried to build `master` and here are the results. Still hitting the same issue, not sure what am I missing out here.

```
PS C:\> git clone git@github.com:grpc/grpc.git
Cloning into 'grpc'...
Enter passphrase for key '/c/Users/nvadapalli/.ssh/id_rsa':
remote: Enumerating objects: 72, done.
remote: Counting objects: 100% (72/72), done.
remote: Compressing objects: 100% (47/47), done.
remote: Total 496937 (delta 43), reused 43 (delta 25), pack-reused 496865
Receiving objects: 100% (496937/496937), 234.53 MiB | 27.31 MiB/s, done.
Resolving deltas: 100% (389784/389784), done.
Updating files: 100% (10483/10483), done.

PS C:\> cd grpc
PS C:\grpc> C:\work\Python1900\python.exe -m virtualenv venv
  No LICENSE.txt / LICENSE found in source
New python executable in C:\grpc\venv\Scripts\python.exe
Installing setuptools, pip, wheel...
done.

PS C:\grpc> . .\venv\Scripts\activate
(venv) PS C:\grpc> pip install -r requirements.txt
DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
Looking in indexes: https://pypi.apps.corp.bardel.ca/simple/, https://pypi.org/simple/
Collecting coverage>=4.0
  Using cached coverage-5.0.3-cp27-cp27m-win_amd64.whl (203 kB)
Collecting cython>=0.29.8
  Using cached Cython-0.29.14-cp27-cp27m-win_amd64.whl (1.7 MB)
Collecting enum34>=1.0.4
  Using cached enum34-1.1.6-py2-none-any.whl (12 kB)
Collecting protobuf>=3.5.0.post1
  Downloading protobuf-3.11.3-py2.py3-none-any.whl (434 kB)
     |UUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUU| 434 kB 7.9 MB/s
Collecting six>=1.10
  Using cached six-1.14.0-py2.py3-none-any.whl (10 kB)
Requirement already satisfied: wheel>=0.29 in c:\grpc\venv\lib\site-packages (from -r requirements.txt (line 7)) (0.34.2)
Requirement already satisfied: setuptools in c:\grpc\venv\lib\site-packages (from protobuf>=3.5.0.post1->-r requirements.txt (line 5)) (44.0.0)
Installing collected packages: coverage, cython, enum34, six, protobuf
Successfully installed coverage-5.0.3 cython-0.29.14 enum34-1.1.6 protobuf-3.11.3 six-1.14.0


(venv) PS C:\grpc> python setup.py bdist_wheel
Cython-generated files are missing...
running bdist_wheel
running build
running build_py
running build_project_metadata
creating python_build
creating python_build\lib.win-amd64-2.7
creating python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_auth.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_channel.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_common.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_compression.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_grpcio_metadata.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_interceptor.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_plugin_wrapping.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_server.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_utilities.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\__init__.py -> python_build\lib.win-amd64-2.7\grpc
creating python_build\lib.win-amd64-2.7\grpc\beta
copying src\python\grpcio\grpc\beta\implementations.py -> python_build\lib.win-amd64-2.7\grpc\beta
copying src\python\grpcio\grpc\beta\interfaces.py -> python_build\lib.win-amd64-2.7\grpc\beta
copying src\python\grpcio\grpc\beta\utilities.py -> python_build\lib.win-amd64-2.7\grpc\beta
copying src\python\grpcio\grpc\beta\_client_adaptations.py -> python_build\lib.win-amd64-2.7\grpc\beta
copying src\python\grpcio\grpc\beta\_metadata.py -> python_build\lib.win-amd64-2.7\grpc\beta
copying src\python\grpcio\grpc\beta\_server_adaptations.py -> python_build\lib.win-amd64-2.7\grpc\beta
copying src\python\grpcio\grpc\beta\__init__.py -> python_build\lib.win-amd64-2.7\grpc\beta
creating python_build\lib.win-amd64-2.7\grpc\experimental
copying src\python\grpcio\grpc\experimental\gevent.py -> python_build\lib.win-amd64-2.7\grpc\experimental
copying src\python\grpcio\grpc\experimental\session_cache.py -> python_build\lib.win-amd64-2.7\grpc\experimental
copying src\python\grpcio\grpc\experimental\__init__.py -> python_build\lib.win-amd64-2.7\grpc\experimental
creating python_build\lib.win-amd64-2.7\grpc\framework
copying src\python\grpcio\grpc\framework\__init__.py -> python_build\lib.win-amd64-2.7\grpc\framework
creating python_build\lib.win-amd64-2.7\grpc\_cython
copying src\python\grpcio\grpc\_cython\__init__.py -> python_build\lib.win-amd64-2.7\grpc\_cython
creating python_build\lib.win-amd64-2.7\grpc\experimental\aio
copying src\python\grpcio\grpc\experimental\aio\_base_call.py -> python_build\lib.win-amd64-2.7\grpc\experimental\aio
copying src\python\grpcio\grpc\experimental\aio\_call.py -> python_build\lib.win-amd64-2.7\grpc\experimental\aio
copying src\python\grpcio\grpc\experimental\aio\_channel.py -> python_build\lib.win-amd64-2.7\grpc\experimental\aio
copying src\python\grpcio\grpc\experimental\aio\_interceptor.py -> python_build\lib.win-amd64-2.7\grpc\experimental\aio
copying src\python\grpcio\grpc\experimental\aio\_server.py -> python_build\lib.win-amd64-2.7\grpc\experimental\aio
copying src\python\grpcio\grpc\experimental\aio\_typing.py -> python_build\lib.win-amd64-2.7\grpc\experimental\aio
copying src\python\grpcio\grpc\experimental\aio\_utils.py -> python_build\lib.win-amd64-2.7\grpc\experimental\aio
copying src\python\grpcio\grpc\experimental\aio\__init__.py -> python_build\lib.win-amd64-2.7\grpc\experimental\aio
creating python_build\lib.win-amd64-2.7\grpc\framework\common
copying src\python\grpcio\grpc\framework\common\cardinality.py -> python_build\lib.win-amd64-2.7\grpc\framework\common
copying src\python\grpcio\grpc\framework\common\style.py -> python_build\lib.win-amd64-2.7\grpc\framework\common
copying src\python\grpcio\grpc\framework\common\__init__.py -> python_build\lib.win-amd64-2.7\grpc\framework\common
creating python_build\lib.win-amd64-2.7\grpc\framework\foundation
copying src\python\grpcio\grpc\framework\foundation\abandonment.py -> python_build\lib.win-amd64-2.7\grpc\framework\foundation
copying src\python\grpcio\grpc\framework\foundation\callable_util.py -> python_build\lib.win-amd64-2.7\grpc\framework\foundation
copying src\python\grpcio\grpc\framework\foundation\future.py -> python_build\lib.win-amd64-2.7\grpc\framework\foundation
copying src\python\grpcio\grpc\framework\foundation\logging_pool.py -> python_build\lib.win-amd64-2.7\grpc\framework\foundation
copying src\python\grpcio\grpc\framework\foundation\stream.py -> python_build\lib.win-amd64-2.7\grpc\framework\foundation
copying src\python\grpcio\grpc\framework\foundation\stream_util.py -> python_build\lib.win-amd64-2.7\grpc\framework\foundation
copying src\python\grpcio\grpc\framework\foundation\__init__.py -> python_build\lib.win-amd64-2.7\grpc\framework\foundation
creating python_build\lib.win-amd64-2.7\grpc\framework\interfaces
copying src\python\grpcio\grpc\framework\interfaces\__init__.py -> python_build\lib.win-amd64-2.7\grpc\framework\interfaces
creating python_build\lib.win-amd64-2.7\grpc\framework\interfaces\base
copying src\python\grpcio\grpc\framework\interfaces\base\base.py -> python_build\lib.win-amd64-2.7\grpc\framework\interfaces\base
copying src\python\grpcio\grpc\framework\interfaces\base\utilities.py -> python_build\lib.win-amd64-2.7\grpc\framework\interfaces\base
copying src\python\grpcio\grpc\framework\interfaces\base\__init__.py -> python_build\lib.win-amd64-2.7\grpc\framework\interfaces\base
creating python_build\lib.win-amd64-2.7\grpc\framework\interfaces\face
copying src\python\grpcio\grpc\framework\interfaces\face\face.py -> python_build\lib.win-amd64-2.7\grpc\framework\interfaces\face
copying src\python\grpcio\grpc\framework\interfaces\face\utilities.py -> python_build\lib.win-amd64-2.7\grpc\framework\interfaces\face
copying src\python\grpcio\grpc\framework\interfaces\face\__init__.py -> python_build\lib.win-amd64-2.7\grpc\framework\interfaces\face
creating python_build\lib.win-amd64-2.7\grpc\_cython\_cygrpc
copying src\python\grpcio\grpc\_cython\_cygrpc\__init__.py -> python_build\lib.win-amd64-2.7\grpc\_cython\_cygrpc
creating python_build\lib.win-amd64-2.7\grpc\_cython\_credentials
copying src\python\grpcio\grpc\_cython\_credentials\roots.pem -> python_build\lib.win-amd64-2.7\grpc\_cython\_credentials
warning: BuildPy: byte-compiling is disabled, skipping.

running build_ext
error: [Error 2] The system cannot find the file specified
(venv) PS C:\grpc> git status
On branch master
Your branch is up to date with 'origin/master'.
```

**Comment by narenandu** [comment]
Btw.. thought attaching the built `python-2.7.11` from the source which was collected in to  `Python1900` (this was referenced in my last comment)
[Python1900.zip](https://github.com/grpc/grpc/files/4150796/Python1900.zip)

**Comment by lidizheng** [comment]
@narenandu Did you update the submodules?

```
git submodule update --init
```

**Comment by narenandu** [comment]
I did that in my original setup. Didn't update submodules after cloning it fresh. Sorry missed that one in the sequence of steps provided before, but here it is after runing the update
```
(venv) PS C:\grpc> git submodule update --init
Submodule 'third_party/abseil-cpp' (https://github.com/abseil/abseil-cpp) registered for path 'third_party/abseil-cpp'
Submodule 'third_party/benchmark' (https://github.com/google/benchmark) registered for path 'third_party/benchmark'
Submodule 'third_party/bloaty' (https://github.com/google/bloaty.git) registered for path 'third_party/bloaty'
Submodule 'third_party/boringssl-with-bazel' (https://github.com/google/boringssl.git) registered for path 'third_party/boringssl-with-bazel'
Submodule 'third_party/cares/cares' (https://github.com/c-ares/c-ares.git) registered for path 'third_party/cares/cares'
Submodule 'third_party/envoy-api' (https://github.com/envoyproxy/data-plane-api.git) registered for path 'third_party/envoy-api'
Submodule 'third_party/gflags' (https://github.com/gflags/gflags.git) registered for path 'third_party/gflags'
Submodule 'third_party/googleapis' (https://github.com/googleapis/googleapis.git) registered for path 'third_party/googleapis'
Submodule 'third_party/googletest' (https://github.com/google/googletest.git) registered for path 'third_party/googletest'
Submodule 'third_party/libuv' (https://github.com/libuv/libuv.git) registered for path 'third_party/libuv'
Submodule 'third_party/protobuf' (https://github.com/google/protobuf.git) registered for path 'third_party/protobuf'
Submodule 'third_party/protoc-gen-validate' (https://github.com/envoyproxy/protoc-gen-validate.git) registered for path 'third_party/protoc-gen-validate'
Submodule 'third_party/udpa' (https://github.com/cncf/udpa.git) registered for path 'third_party/udpa'
Submodule 'third_party/zlib' (https://github.com/madler/zlib) registered for path 'third_party/zlib'
Cloning into 'C:/grpc/third_party/abseil-cpp'...
Cloning into 'C:/grpc/third_party/benchmark'...
Cloning into 'C:/grpc/third_party/bloaty'...
Cloning into 'C:/grpc/third_party/boringssl-with-bazel'...
Cloning into 'C:/grpc/third_party/cares/cares'...
Cloning into 'C:/grpc/third_party/envoy-api'...
Cloning into 'C:/grpc/third_party/gflags'...
Cloning into 'C:/grpc/third_party/googleapis'...
Cloning into 'C:/grpc/third_party/googletest'...
Cloning into 'C:/grpc/third_party/libuv'...
Cloning into 'C:/grpc/third_party/protobuf'...
Cloning into 'C:/grpc/third_party/protoc-gen-validate'...
Cloning into 'C:/grpc/third_party/udpa'...
Cloning into 'C:/grpc/third_party/zlib'...
Submodule path 'third_party/abseil-cpp': checked out '37dd2562ec830d547a1524bb306be313ac3f2556'
Submodule path 'third_party/benchmark': checked out '090faecb454fbd6e6e17a75ef8146acb037118d4'
Submodule path 'third_party/bloaty': checked out '73594cde8c9a52a102c4341c244c833aa61b9c06'
Submodule path 'third_party/boringssl-with-bazel': checked out '1c2769383f027befac5b75b6cedd25daf3bf4dcf'
Submodule path 'third_party/cares/cares': checked out 'e982924acee7f7313b4baa4ee5ec000c5e373c30'
Submodule path 'third_party/envoy-api': checked out 'c83ed7ea9eb5fb3b93d1ad52b59750f1958b8bde'
Submodule path 'third_party/gflags': checked out '28f50e0fed19872e0fd50dd23ce2ee8cd759338e'
Submodule path 'third_party/googleapis': checked out '80ed4d0bbf65d57cc267dfc63bd2584557f11f9b'
Submodule path 'third_party/googletest': checked out 'c9ccac7cb7345901884aabf5d1a786cfa6e2f397'
Submodule path 'third_party/libuv': checked out '15ae750151ac9341e5945eb38f8982d59fb99201'
Submodule path 'third_party/protobuf': checked out 'fe1790ca0df67173702f70d5646b82f48f412b99'
Submodule path 'third_party/protoc-gen-validate': checked out 'e143189bf6f37b3957fb31743df6a1bcf4a8c685'
Submodule path 'third_party/udpa': checked out '94324803a497c8f76dbc78df393ef629d3a9f3c3'
Submodule path 'third_party/zlib': checked out 'cacf7f1d4e3d44d871b605da3b647f07d718623f'
(venv) PS C:\grpc> python setup.py bdist_wheel
Cython-generated files are missing...
running bdist_wheel
running build
running build_py
running build_project_metadata
copying src\python\grpcio\grpc\_grpcio_metadata.py -> python_build\lib.win-amd64-2.7\grpc
copying src\python\grpcio\grpc\_cython\_credentials\roots.pem -> python_build\lib.win-amd64-2.7\grpc\_cython\_credentials
warning: BuildPy: byte-compiling is disabled, skipping.

running build_ext
error: [Error 2] The system cannot find the file specified
```

**Comment by lidizheng** [comment]
@narenandu Is there a reason you need to manually compile gRPC Python? It's not easy to get it right. Do you think you can use the latest Windows binary wheels through `pip install -U grpcio`?

Did you have Visual Studio installed? Can you try to figure out which file is missing?

**Comment by narenandu** [comment]
@lidizheng Reason why I have to manully compile `grpc` python is because of its usage in a Digitial Content Creation App (DCC) called [Autodesk Maya](https://www.autodesk.ca/en/products/maya/overview). Maya comes with its bundled Python interpreter and its own set of `site-packages`. This fact points us back to why we need to compile `Python` from its source: https://clamdragon3d.com/blog/2018/3/22/python-virtualenvs-and-maya

Idea was to compile Python first and then use it to build `whl` file, so that it is compatible with `Maya`'s Python interpreter. Otherwise, we keep hitting something like the following

```
.venv\Lib\site-packages\grpc\__init__.py:23: in <module>
    from grpc._cython import cygrpc as _cygrpc
E   ImportError: DLL load failed: The specified module could not be found.
```
I do have VS2015 installed and am trying to figure out a way to turn up the verbosity to `python setup.py bdist_wheel` by passing a `-v` flag. But can't seem to get more information out of it. I will keep trying to find the root cause. But thanks for the help so far !

**Comment by johnwog** [comment]
@narenandu I resolved the issue using the patches provided by the grpcio conda-recipes repo. might also help your case too

**Comment by narenandu** [comment]
@johnwog will check that: https://github.com/conda-forge/grpcio-feedstock/blob/master/recipe/bld.bat is this the repo which you are pointing out?

**Comment by johnwog** [comment]
yes there are .patch files in that folder too. they might do the trick for
you.

**Comment by narenandu** [comment]
@johnwog / @gnossen 

I am happy to open a new issue to discuss this further, let me know. But here are the details of my debugging so far. Custom built Python has a `Lib/distutils/cygwinccompiler.py` that has the following method
```
```
